arch/risc-v/src/common/riscv_mmu.h

```diff
@@ -0,0 +1,306 @@
+#define PTE_X                   (1 << 3) /* Page is executable */
+#define PTE_U                   (1 << 4) /* Page is a user mode page */
+#define PTE_G                   (1 << 5) /* Page is a global mapping */
+#define PTE_A                   (1 << 6) /* Page has been accessed */
+#define PTE_D                   (1 << 7) /* Page is dirty */
+
+/* Check if leaf PTE entry or not (if X/W/R are set it is) */
+
+#define PTE_LEAF_MASK           (7 << 1)
+
+/* Flags for user page tables */
+
+#define MMU_UPGT_FLAGS          (0)
+
+/* Flags for user FLASH (RX) and user RAM (RW) */
+
+#define MMU_UTEXT_FLAGS         (PTE_R | PTE_X | PTE_U)
+#define MMU_UDATA_FLAGS         (PTE_R | PTE_W | PTE_U)
+
+/* SvX definitions, only Sv39 is currently supported, but it should be
+ * trivial to extend the driver to support other SvX implementations
+ *
+ * Sv39 has:
+ * - 4K page size
+ * - 3 page table levels
+ * - 9-bit VPN width
+ */
+
+#ifdef CONFIG_ARCH_MMU_TYPE_SV39
+#define RV_MMU_PTE_PADDR_SHIFT  (10)
+#define RV_MMU_PTE_PPN_MASK     ((1 << RV_MMU_PTE_PADDR_SHIFT) - 1)
+#define RV_MMU_PTE_PPN_SHIFT    (2)
+#define RV_MMU_VPN_WIDTH        (9)
+#define RV_MMU_VPN_MASK         ((1 << RV_MMU_VPN_WIDTH) - 1)
+#define RV_MMU_VADDR_SHIFT(_n)  (RV_MMU_PAGE_SHIFT + RV_MMU_VPN_WIDTH * \
+                                 (RV_MMU_PT_LEVELS - (_n)))
+#define RV_MMU_SATP_MODE        (SATP_MODE_SV39)
+#define RV_MMU_L1_PAGE_SIZE     (0x40000000) /* 1G */
+#define RV_MMU_L2_PAGE_SIZE     (0x200000)   /* 2M */
+#define RV_MMU_L3_PAGE_SIZE     (0x1000)     /* 4K */
+#else
+#error "Unsupported RISC-V MMU implementation selected"
+#endif /* CONFIG_ARCH_MMU_TYPE_SV39 */
+
+/****************************************************************************
+ * Name: satp_reg
+ *
+ * Description:
+ *   Utility function to build satp register value for input parameters
+ *
+ * Input Parameters:
+ *   pgbase - The physical base address of the translation table base
+ *   asid - Address space identifier. This can be used to identify different
+ *     address spaces. It is not necessary to use this, nor is it necessary
+ *     for the RISC-V implementation to implement such bits. This means in
+ *     practice that the value should not be used in this generic driver.
+ *
+ ****************************************************************************/
+
+static inline uint64_t satp_reg(uint64_t pgbase, uint16_t asid)
+{
+  uint64_t reg;
+  reg  = ((RV_MMU_SATP_MODE << SATP_MODE_SHIFT) & SATP_MODE_MASK);
+  reg |= (((uint64_t)asid << SATP_ASID_SHIFT) & SATP_ASID_MASK);
+  reg |= ((SATP_ADDR_TO_PPN(pgbase) << SATP_PPN_SHIFT) & SATP_PPN_MASK);
+  return reg;
+}
+
+/****************************************************************************
+ * Name: mmu_write_satp
+ *
+ * Description:
+ *   Write satp
+ *
+ * Input Parameters:
+ *   reg - satp value
+ *
+ ****************************************************************************/
+
+static inline void mmu_write_satp(uint64_t reg)
```

Comment:
Use uintptr_t to represent register width value.

Reply:
uintptr_t does not quite represent a CPU registers width. Quote from the C standard is as follows:

The following type designates an unsigned integer type with the property that any valid
pointer to void can be converted to this type, then converted back to pointer to void,
and the result will compare equal to the original pointer:

uintptr_t

I can create an abstract register type (dependent on __riscv_xlen which tells the architecture specific register size) if that is acceptable ?

Reply:
Pointer type represent to register with usually on NuttX at least:
```
/* Integer types capable of holding object pointers
 * As a general rule, the size of size_t should be the same as the size of
 * uintptr_t: 32-bits on a machine with 32-bit addressing but 64-bits on a
 * machine with 64-bit addressing.
 */

typedef _ssize_t            intptr_t;
typedef _size_t             uintptr_t;
```

Reply:
Ok, will change.

Should I change the address parameters given to e.g. 

void mmu_ln_setentry(uint32_t ptlevel, uint64_t lnvaddr, uint64_t paddr,
                     uint64_t vaddr, uint32_t mmuflags);

To uintptr type as well ?

Reply:
Yes, if this function will support Sv32 also, all address parameters should change to uintptr_t.

Reply:
Thanks, will re-factor the arguments